# День пятый DWthon (завершающий) 👋 - выйди за рамки стандартного мышления!


### Цель состоит в том, чтобы узнать, что такое так называемый «aha moment»!
Мы уже сейчас подошли к очень интересному термину - «ага момент». Что это такое? Простым языком - это момент, когда Тебя "осеняет" вырывается слово ... «ага» 😁. 

Можно долго говорить об «ага моменте», лучше приведу сразу несколько примеров.


### Facebook (7 друзей за 10 дней)
Если пользователь добавит не менее **7 друзей** в течение **первых 10 дней** после регистрации, то этот пользователь будет более вовлечен и останется с Facebook дольше.

### Dropbox (1 файл, 1 папка, 1 устройство)
Если пользователь добавляет хотя бы один файл, создаст одну папку и синхронизируется хотя бы с одним устройством, то этот пользователь останется дольше.

### LinkedIn - X подключений за Y дней
Похоже с историей в Facebook.

Это простые правила, которые могут описать самое важное, что должно произойти в компании для ее роста. 

Это все ага-моменты, которые были найдены в разных компаниях. Другими словами - это явное применение правила 80/20. 
Только обрати внимание, что эти "ага моменты" звучат очень просто, когда уже найдены. А вот как их найти? Ведь возможных комбинаций этих условий может быть очень много. Нужен какой-то инструмент, который поможет сделать это эффективнее.  В этом нам может помочь машинное обучение (англ. `Machine Learning`).

### ☝️ Прежде, чем мы начнем, важно (!)
Каждый день DWthon проводится специальный опрос. Пожалуйста, пройди его, прежде, чем приступить к обучению. В конце концов - это может быть приятным сюрпризом :) Это займет у тебя **только** 1 минуту.


##  👉👉 **[Заполни анкету](https://bit.ly/2RElpDI)** 👈👈 и измерь дельту своего развития 💪

### Где задавать вопросы ❓ 
Если ты столкнулся(лась) с трудностями при выполнении задач из этого дня, обязательно размести свой вопрос в чате [#dwthon_day5](https://bit.ly/3z2qO8p)

Не забудь подробно указать, с чем у Тебя возникла проблема. Лучше всего выложить скрин со своим кодом или обнаруженной ошибкой и написать, к какой задаче это относится :)

In [1]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/pNe7Ox85_II" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

### Шаг за шагом
Я также записал для Тебя "пошаговое" видео. В видео ниже Ты найдешь именно то, что в этом ноутбуке, только со всеми объяснениями, чтобы облегчить Твою работу и понять проблемы и задачи :) Посмотри, нужна ли Тебе такая дополнительная поддержка.

In [2]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/PVdgThGiBhM" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Библиотеки

Сегодня нам понадобятся библиотеки

- `pandas` - работа с данными
- `import numpy as np` - полезные функции (кстати, `pandas` внутри использует много `numpy`)
- `DecisionTreeClassifier` - дерево решений 
- `sklearn.model_selection` - проверка модели
- `xgboost` - более продвинутая модель  
- `eli5` - интерпретирование моделей, чтобы увидеть важность этих функций
- `from collections import Counter` это будет полезно для удобного подсчета количества повторов (например, сколько раз покупатель покупал продукт x).
- `gc` - сборщик "мусора"

In [3]:
import pandas as pd
import numpy as np

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
import xgboost as xgb
import eli5 
from collections import Counter

import gc

## Загрузим данные

In [4]:
df = pd.read_hdf("../input/data.h5")
print(df.shape)
df.sample(5)

(820906, 9)


order_id  customer_id  product_id  quantity  price_unit  price_total  \
688904     45333         2105         409        12         165         1980   
141770     10014         2705        1984        24         125         3000   
765286     49929           40        2683         4         795         3180   
552319     35798         5156        2729         3         125          375   
149334     10653         1467        2572         3         325          975   

        country_id          order_date  is_canceled  
688904           0 2010-09-28 09:12:00        False  
141770          20 2011-05-23 09:43:00        False  
765286           0 2010-11-11 12:19:00        False  
552319          11 2010-05-10 16:44:00        False  
149334           0 2011-06-02 11:33:00        False

## Подготавливаем данные


Это код из предыдущего урока. Мы просто преобразуем код в функцию, чтобы нам было легче экспериментировать позже.

In [5]:
df_customers = (
    df[ ["price_total", "customer_id"] ]
    .groupby("customer_id")
    .agg("sum")
    .reset_index()
    .sort_values(by="price_total", ascending=False)
    .rename(columns={"price_total": "customer_price_total"})
)


df_customers["cumsum"] = df_customers["customer_price_total"].cumsum()
value_80prc = int(df["price_total"].sum() * 0.8)
df_customers["most_revenue_customer"] = df_customers["cumsum"] < value_80prc


top_customers = set(df_customers[ df_customers["most_revenue_customer"] ]["customer_id"].unique())

del df_customers
gc.collect()

def feature_engineering(df):
    df_customers = (
        df
        .groupby("customer_id")
        .agg(
            count_orders=("order_id", lambda x: len(set(x))),
            count_unq_products=("product_id", lambda x: len(set(x))),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        ).reset_index()
    )
    
    
    
    return df_customers


def get_feats(df_customers, black_list=["most_revenue_customer"]):
    feats = list(df_customers.select_dtypes([np.number, bool]).columns)
    return [x for x in feats if x not in black_list]

def get_X_y(df_customers, top_customers, feats):
    df_customers["most_revenue_customer"] = df_customers["customer_id"].map(lambda x: x in top_customers)
    
    X = df_customers[feats].values
    y = df_customers["most_revenue_customer"].values
    
    return X, y


def train_and_get_scores(model, X, y, scoring="accuracy", cv=5):

    scores = cross_val_score(model, X, y, scoring=scoring, cv=cv)
    return np.mean(scores), np.std(scores)

## Модель 1

👇 Именно то, что мы сделали недавно.

In [6]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers)
X, y = get_X_y(df_customers, top_customers, feats)
model = DecisionTreeClassifier(max_depth=3)

train_and_get_scores(model, X, y)

(0.9996598639455783, 0.0004165798882284457)

У нас результат 99,9%. Это подозрительно хороший результат.

Можешь быть счастлив, но... на самом деле я расскажу Тебе об одном механизме размышления, который я стараюсь использовать регулярно. Я называю это маятником. Хотя думаю, что этот метод был выдуман давным давно (и возможно, даже имеет некоторое название), но это не так важно. Ведь в реальной жизни важнее уметь применять, чем делать вид "что все знаешь" и при этом используешь умные слова, но понятия не имеешь - как это действительно работает. Согласен(а)?

Идея состоит в том, чтобы сначала подходить оптимистично и даже можно сильно упрощать реальность. Затем, когда у нас есть хороший результат, мы «переобуваемся». Обычно это звучит в негативном ключе, но здесь смысл в другом. Посмотреть на ситуацию с другой стороны. Мы играем роль активного «критика», который пытается сделать все, чтобы доказать бессмысленность того, что мы сами же и сделали. И переключение от оптимиста в сторону критика повторяем снова и снова. Делаем несколько итераций :). 
Это порой утомляющий процесс, когда выкладываешься на 100% и на оба фронта. Часто требуются даже разные компетенции - именно поэтому хорошо, когда в одной команде есть люди с одного и со второго лагеря.

Но самое главное, что, переключаясь между ролями, можно очень быстро развиваться и прощупывать возможности, а также минимизировать риски. В результате потихоньку, но решительно двигаешься вперед!

## 🧠 Включим критическое мышление

На данный момент мы сделали первый шаг, который является оптимистичным, и получили результат 99%. Очень хорошо, но ... Проблема в том, что `price_unit_total` слишком сильная характеристика. Таким образом, эта функция по сути напрямую сообщает, принадлежит ли покупатель к сегменту `top_customers` или нет. Как именно? Давай вспомним, что значит эта цифра? Это общее количество денег, которые потратил клиент в нашем интернет магазине (который анализируем). Существует некоторая граница, пересекая которую, начинаешь принадлежать к сегменту `top_customers`. Говоря простым языком, каждый, кто потратил миллион (или даже гораздо меньше) становится "любимчиком" среди клиентов. Соответственно, наш алгоритм (даже относительно простой, как дерево решений) найдет это автоматически. Не веришь? Давай проверять.


Давай быстро посмотрим что там внутри - для этого используем `eli5` (eli5 - это сокращение от `Explain Like I'm 5` т.е. объясни мне как пятилетнему ребенку, что подразумевает, что объяснение должно быть очень простым).

In [7]:
model = DecisionTreeClassifier(max_depth=3)
model.fit(X, y)
eli5.show_weights(model, feature_names=feats)

## Что это значит и как это интерпретировать?

Сразу видно таблицу в двумя колонками:
- Weight - вес важности признака
- Feature - название признака

Видим, что `sum_price_total` по сути сделал "всю работу" и модель быстро "уловила" важность этого признака. Что значит "уловила"? Чуть ниже (под таблицей) есть дерево решений (три прямоугольника и две стрелки, видишь? 👀). 

В корне дерева (т.е. самый верхний прямоугольник,  кстати, здесь наоборот - корень дерева вверху 😉) видно 4 строки:
- `sum_price_total <= 263595.0`
- `gini = 0.349`
- `samples = 100.0%`
- `value = [0.775, 0.225]`

Самое интересное - это первая строка `sum_price_total <= 263595.0` - это есть условие. Дословно это значит, что если клиент потратил **£2635,95** тогда он присоединяется в клуб `most_revenue_customer`. Благодаря этому условию, задача решается при помощи одного "удара" (разделения). Но это задача достаточно простая, поэтому алгоритм быстро нашел решение.


Кстати, если интересно, то можно немного усложнить жизнь для алгоритма и удалить признак `"sum_price_total"`, а точнее говоря - спрятать этот признак, чтобы его не было в X. Давай быстро проведем эксперимент.

In [8]:
black_list = ["most_revenue_customer", "sum_price_total"]
feats = [x for x in feats if x not in black_list]

X = df_customers[feats].values
y = df_customers["most_revenue_customer"].values


model = DecisionTreeClassifier(max_depth=3)
model.fit(X, y)
eli5.show_weights(model, feature_names=feats)

Теперь видно более "ветвистое" и сложное дерево. Также видно, что сейчас доминирует признак `sum_quantity`, который считает количество штук среди всех заказов для каждого клиента.

Можем даже немного вместе проследить, как появились условия:
- `sum_quantity <= 1534.5`
- `count_orders <= 10.5`
- `um_price_unit <= 31075.5`
- ...


Кстати, а давай проверим как "просело" качество модели, когда спрятали признак `sum_price_total`?

In [9]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers, top_customers, feats)
model = DecisionTreeClassifier(max_depth=3)

train_and_get_scores(model, X, y)

(0.7863585178752353, 0.26358237322275296)

Как видишь, качество модели упало до 78% (было 99%). С одной стороны, грустно 😂, но с другой стороны, мы используем модель машинного обучения, чтобы найти менее очевидные зависимости. Потому что очевидные вещи легко найти самостоятельно на пальцах или калькуляторе 🤦‍♂️.


### Оптимистичное мышление
А теперь вернемся к оптимисту. Ты помнишь, в чем принцип маятника? Мы сомневаемся или наоборот - полны оптимизма, и так вперед и назад - по кругу;)

Сейчас мы перейдем на более трудную модель и это должно улучшить качество модели (улавливать более сложные взаимосвязи). Давай использовать алгоритмы из семейства "бустинг" - а именно `xgboost`. Это название может тебе ни о чем не говорить, и на данном этапе - это нормально 👌. Просто запомни, что - это проверенный в бою алгоритм, который также ведет себя уверенно и на "продакшене" (англ. `production`).

## XGBoost
Обрати внимание, что, по сравнению с предыдущим кодом, изменилась только эта строка: `model = xgb.XGBClassifier(**xgb_params)` и добавили параметры для нашей модели в пару строк (чтобы не делать одну длинную т.к. это трудно читается).

In [10]:
df_customers = feature_engineering(df)
feats = get_feats(df_customers, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers, top_customers, feats)

xgb_params = dict(
    max_depth=5, 
    n_estimators=50, 
    learning_rate=0.3, 
    objective="binary:logistic", 
    eval_metric="error",
    random_state=0)
model = xgb.XGBClassifier(**xgb_params)

train_and_get_scores(model, X, y)

(0.8106769431176726, 0.08696513375768444)

## У нас результат 81% 🥳

Попробуем сделать первую интерпретацию модели и посмотрим, что она считает важным.

In [11]:
model.fit(X, y)

eli5.show_weights(model, feature_names=feats)

В данном случае у нас уже не одно дерево внутри, а целых 50 штук (за это отвечает параметр `n_estimators=50`).  Кстати, может быть гораздо больше: 100 или даже больше 1000, но пока что нам это не нужно. Давай сделаем первые наблюдения и выводы.

### Предварительные выводы:
#### TOP3 
- `sum_quantity` звучит как самая **важная** особенность* (осторожно, ниже комментарий к звездочке)
- `count_orders`
- `sum_price_unit`

Похоже, что функция `count_unq_countries` мало интересна.


☝️ * Нужно быть осторожным с тем, что является «важным», потому что опыт показывает, что это может быть довольно изменчивым, и нужно научиться ориентироваться в этой изменчивости. Если бы все было так просто, то этот процесс уже давно был бы на 100% автоматизирован. 


## Следующий шаг

Теперь мы можем немного порасcуждать над этими признаками, но это не имеет большого значения. Эти особенности сами по себе очевидны. Даже если мы сделаем правильные выводы (что не факт), то какая будет ценность? Нужно все-таки использовать машинное обучение (желательно находить менее очевидные зависимости, которые вызывают восторг и удивление!)

Давай сгенерируем больше признаков и пусть модель сама найдет то, на что стоит обратить внимание :)

## Продукты
Мы можем добавить отдельный столбец для каждого продукта и добавить значения в этот столбец -  сколько продуктов уже было куплено (то есть сумму, например, Саша купил 5 упаковок молока). Таким образом, мы хотим выяснить, какие продукты «самые интересные», на которые стоит обратить внимание.

☝️ Конечно, мы можем проверить каждый продукт (товар) вручную, но помни, что у нас их более 3.8 тыс. Делать проверку вручную - это  очень утомительно, занимает кучу времени да еще и дорогим процессом является (зарплаты нужно добавить и т.д). Давай лучше думать по принципу 80/20 и использовать правильные решения, например, машинное обучение, оно будет быстрее, надежнее и эффективнее.

Только сначала нам нужно подготовить признаки. И желательно, сделать их гораздо больше (чтобы было где "разгуляться" алгоритму). Давай изменим нашу функцию `feature_engineering`. 

In [12]:
def feature_engineering(df):
    
    def counter(vals):
        cntr = Counter()
        cntr.update(vals)
        return cntr
    
    df_customers = (
        df
        .groupby("customer_id")
        .agg(
            count_orders=("order_id", lambda x: len(set(x))),
            count_unq_products=("product_id", lambda x: len(set(x))),
            count_by_products=("product_id", lambda x:  counter(x) ),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        ).reset_index()
    )
    
    
    
    return df_customers


df_customers  = feature_engineering(df)
df_customers.head()

customer_id  count_orders  count_unq_products  \
0            0           159                  93   
1            1            40                 184   
2            2            30                 139   
3            3             9                  40   
4            4            14                  11   

                                   count_by_products  sum_quantity  \
0  {0: 89, 1: 73, 2: 69, 3: 68, 4: 32, 5: 36, 6: ...         22976   
1  {8: 8, 9: 4, 10: 4, 11: 4, 12: 3, 13: 15, 14: ...          3514   
2  {25: 9, 26: 7, 27: 7, 28: 3, 29: 4, 30: 10, 31...          8971   
3  {44: 7, 1781: 1, 1179: 1, 112: 1, 508: 3, 581:...          1108   
4  {54: 14, 1496: 2, 2288: 1, 2315: 1, 3432: 1, 4...           390   

   sum_price_unit  sum_price_total  count_unq_countries  prob_canceled  
0          895827          5746671                    1       0.010966  
1          205962           910113                    1       0.097727  
2           95500          1337464                    1       0.009050  
3           18626           226381                    1       0.000000  
4           20028           346978                    1       0.307692

Появился новый столбец `count_by_products`. Внутри есть словарь. Посмотрим внимательнее на этот словарь 👀.

In [13]:
df_customers["count_by_products"]

0       {0: 89, 1: 73, 2: 69, 3: 68, 4: 32, 5: 36, 6: ...
1       {8: 8, 9: 4, 10: 4, 11: 4, 12: 3, 13: 15, 14: ...
2       {25: 9, 26: 7, 27: 7, 28: 3, 29: 4, 30: 10, 31...
3       {44: 7, 1781: 1, 1179: 1, 112: 1, 508: 3, 581:...
4       {54: 14, 1496: 2, 2288: 1, 2315: 1, 3432: 1, 4...
                              ...                        
5874    {1709: 1, 409: 1, 170: 1, 662: 1, 253: 1, 255:...
5875    {2368: 1, 300: 1, 1385: 1, 2412: 1, 2059: 1, 1...
5876    {1827: 1, 869: 1, 2157: 1, 335: 1, 1931: 1, 21...
5877    {1038: 1, 701: 2, 495: 1, 353: 1, 189: 1, 484:...
5878                                             {694: 2}
Name: count_by_products, Length: 5879, dtype: object

Давай для примера возьмем первую строку и прочитаем вместе - что там. На практике это означает, что `customer_id=0` купил `product_id= 0` - 89 раз, `product_id=1` - 73 раза,` product_id=3` - 68 раз и т.д.


Было бы неплохо, если бы мы распаковали словарь на отдельные столбцы, благо сделать это очень просто, используя `pandas` и пару простых трюков. ⌛️ Но нужно немного подождать пока сервер будет работать.

In [14]:
df_count_products = df_customers["count_by_products"].apply(pd.Series).fillna(-1)
df_count_products.columns = ["product_{}".format(x) for x in df_count_products.columns]

df_count_products.head(5)

product_0  product_1  product_2  product_3  product_4  product_5  \
0       89.0       73.0       69.0       68.0       32.0       36.0   
1        6.0       -1.0        1.0       -1.0       -1.0       -1.0   
2       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   
3       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   
4       -1.0       -1.0       -1.0       -1.0       -1.0       -1.0   

   product_6  product_7  product_8  product_9  ...  product_4216  \
0       34.0       35.0       18.0       -1.0  ...          -1.0   
1       -1.0       -1.0        8.0        4.0  ...          -1.0   
2       -1.0       -1.0       -1.0       -1.0  ...          -1.0   
3       -1.0       -1.0       -1.0       -1.0  ...          -1.0   
4       -1.0       -1.0       -1.0       -1.0  ...          -1.0   

   product_4219  product_4220  product_4222  product_4226  product_4231  \
0          -1.0          -1.0          -1.0          -1.0          -1.0   
1          -1.0          -1.0          -1.0          -1.0          -1.0   
2          -1.0          -1.0          -1.0          -1.0          -1.0   
3          -1.0          -1.0          -1.0          -1.0          -1.0   
4          -1.0          -1.0          -1.0          -1.0          -1.0   

   product_4236  product_4237  product_4238  product_4239  
0          -1.0          -1.0          -1.0          -1.0  
1          -1.0          -1.0          -1.0          -1.0  
2          -1.0          -1.0          -1.0          -1.0  
3          -1.0          -1.0          -1.0          -1.0  
4          -1.0          -1.0          -1.0          -1.0  

[5 rows x 3878 columns]

Теперь у нас есть отдельный `dataframe` с 3878 столбцами. Каждый столбец - это `product_id`. Значение -1 означает отсутствие информации, т.е. покупатель не покупал этот товар, а любое другое число (больше 0) означает -  сколько раз этот товар был куплен (это немного оптимистичное мнение, но пока мы можем предположить, что это в порядке).

☝️ Обрати внимание, что появилась строка для переименования продуктов. `df_count_products.columns = ["product _ {". format(x) for x in df_count_products.columns]` Здесь мы делаем очень простые вещи, вместо простого идентификатора, например 0 или 2, мы добавляем префикс для столбца `product_0` или `product_2`. Это нам сейчас пригодится.

### Следующий шаг
Давай объединим новую табличку `df_count_products` с таблицей `df_customers` и на всех этих данных обучим модель. Чтобы объединить это, мы будем использовать функцию `concat()`. 

Объясняю на пальцах, что мы хотим сделать. Имеется:
- таблица 1 `df_customers.shape` => `(5879, 9)` т.е. 5879 строк и 9 столбцов
- таблица 2 `df_count_products.shape` => `(5879, 3878)` т.е. 5879 строк и 3878 столбцов

Как видишь, количество строк одинаковое. И это, конечно, не случайно, ведь это наше количество клиентов. То, что появилось в таблице 2 - это дополнительная информация для каждого клиента. Теперь нам нужно объединить в одну таблицу все столбцы (признаки) о клиенте, т.е. в результате объединения получается `9+3878=3887` столбцов.

Можно даже так сказать, берем две таблицы и прикладываем их друг к другу, как две книги. Другими словами, количество строк останется прежним, количество столбцов увеличится на 3878 😱. Алгоритму это обязательно "понравится"😉 (конечно, у модели нет собственного мнения, но если бы оно было - то скорее всего - оно было бы такое, либо нам хотелось бы, чтобы оно такое было 🤫), т.к. будет больше признаков и можно будет найти что-то интересное, ну или по крайней мере - будет что оптимизировать.

Хватит слов, объединяем!

In [15]:
df_customers_big = pd.concat([df_customers, df_count_products], axis=1)
df_customers_big.shape

(5879, 3887)

Как и ожидалось, количество строк осталось прежним - 5879 (поскольку количество строк - это клиенты, а новых клиентов не было). Количество столбцов - увеличилось до 3887. Теперь можно сказать, что для каждого покупателя у нас теперь 3887 признаков (характеристик).

## 🤖 Время обучать модель

Для начала проверим, улучшилось ли для нас качество модели.

⌛️⌛️⌛️ Внимание! Данных стало больше (было 9 признаков, а стало 3887), комбинаций стало гораздо больше, поэтому алгоритму нужно время на тренировку модели.

Алгоритм выполняет «простую» и повторяющуюся работу, в то время, как Ты можешь тихонько в стороне мыслить дальше, о чем-то более концептуальном 😉. Привыкай - это будущее, которое к нам уже пришло :).

In [16]:
feats = get_feats(df_customers_big, black_list=["most_revenue_customer", "sum_price_total"])
X, y = get_X_y(df_customers_big, top_customers, feats)

xgb_params = dict(
    max_depth=5, 
    n_estimators=50, 
    learning_rate=0.3, 
    objective="binary:logistic", 
    eval_metric="error",
    random_state=0)
model = xgb.XGBClassifier(**xgb_params)

train_and_get_scores(model, X, y)

KeyboardInterrupt: 

Качество модели улучшилось, у нас уже ~ 85%.

Напомню, что только деревья решений дали нам 77%. XGBoost по базовым характеристикам дал нам точность 81% и теперь, добавив новых признаков, улучшили точность модели до ~ 85%.

Может показаться, что это не много, потому что у нас было 99,9%. Хочу снова такого результата! 😂

Возможно, это плохие новости, но часто модель «останавливается» на отметке 60-70% и тогда приходится очень стараться выжать больше. Ну разве что мы решаем очевидные задачи, но тогда не очень интересно применение машинного обучения (трудно объяснить выгоду). Это как за булочками ездить на Боинге - немного неудобно (хотя летать между континентами - очень даже хорошо). 

Теперь давай обучим еще раз модель, чтобы у нас был доступ к ней напрямую. Выше мы использовали дополнительные функции, которые удобны, но сама модель "теряется" где-то в недрах эксперимента и до нас доходит только информация, например, о качестве (это тоже порой удобно).

Тренируем модель с нуля, чтобы у нас был доступ напрямую к ее состоянию (т.е. конечное решение).

⌛️⌛️⌛️ Внимание! Долго считает.

In [ ]:
model.fit(X, y)

Теперь пришло время посмотреть на важность признаков.

In [ ]:
eli5.show_weights(model, feature_names=feats, top=50)

Что ж, теперь хоть есть на что посмотреть :). Ты можешь увидеть, что важность признаков `product_545` и` product_99` превышает даже важность `count_orders`. Конечно, как я уже сказал, здесь нужно делать выводы осторожно. Однако однозначно то, что `product_id=545` заслуживает нашего внимания. 🥳 Ну что ж - продукт 545 - держись, мы идем тебя проверять :).

In [ ]:
customer_ids_by_product = set(df[ df["product_id"] == 545 ]["customer_id"].unique())
len(customer_ids_by_product )

Этот товар купили не так много клиентов, всего 168. Посмотрим, сколько из них относятся к сегменту `most_revenue_customer`.

In [ ]:
df_customers["most_revenue_customer"] = df_customers["customer_id"].map(lambda x: x in top_customers) 

In [ ]:
df_customers[ df_customers.customer_id.isin(customer_ids_by_product) ]["most_revenue_customer"].mean()

Итак, мы видим, что у нас есть 42% клиентов, которые купили по крайней мере `product_id=545` и принадлежат к сегменту `most_revenue_customer`.

Напомню, что у нас обычно 22% клиентов относятся к сегменту `most_revenue_customer`.

In [ ]:
df_customers["most_revenue_customer"].mean()

Мы можем быстро выдвинуть гипотезу (для проверки) - если покупатель купил `product_id=545`, вероятность того, что он будет принадлежать сегменту `most_revenue_customer`, почти удвоится (с 23% до 42%). Это звучит как "aha moment". Эта гипотеза заслуживает дальнейшего изучения, но если она, действительно, работает, это означает, что стоит просто побудить людей больше покупать `product_id=545`.

#### Давай узнаем, сколько стоит этот товар. Может это какая-то роскошь ⚜️.

In [ ]:
df[ df.product_id == 545 ]["price_unit"].value_counts()

Цена на этот товар варьируется, но чаще всего цена за единицу составляет всего 1,25 или 3,9. Такой интересный вывод был сделан так легко.

## Задание 5.1


Твоя задача - сделать аналогичный анализ для продукта. Попробывать понять, что очень сильно влияет на продукт, что он становится `top_products`. 

Когда уже построишь модель и найдешь что влияет больше всего, сделай выводы... даже можешь попробывать создать гипотезу для нового продукта - что нужно вложить в продукт чтобы попал в TOP за ограниченое время (здесь тоже можешь подумать как это время определить разумно с точки зрения бизнеса).

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть подсказку 👈 </summary>
<p>
    <h3>💡 Подсказки</h3>

1. Подготовь `df_products`
2. Добавь больше признаков. Прежде мы добавили `product_id`, но Ты можешь сделать симметричным или придумать что-то еще (возможно ,` order_id`) или что-то еще, идей много :)
3. Обучи модель.
4. Посмотри, что было самым важным для модели.

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть ответ 👈 </summary>
<p>
    
```python
def get_top_products(df):
    df_products = (
        df[ ["price_total", "product_id"] ]
        .groupby("product_id")
        .agg("sum")
        .reset_index()
        .sort_values(by="price_total", ascending=False)
        .rename(columns={"price_total": "product_price_total"})
    )


    df_products["cumsum"] = df_products["product_price_total"].cumsum()
    value_80prc = int(df["price_total"].sum() * 0.8)
    df_products["most_revenue_product"] = df_products["cumsum"] < value_80prc


    top_products = set(df_products[ df_products["most_revenue_product"] ]["product_id"].unique())

    del df_products
    gc.collect()
    return top_products

def feature_engineering(df):
    
    def counter(vals):
        cntr = Counter()
        cntr.update(vals)
        return cntr
    
    df_products = (
        df
        .groupby("product_id")
        .agg(
            ## HERE YOU CAN PUT SOMETHING MORE
            counter_by_customers=("customer_id", lambda x:  counter(x) ),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        ).reset_index()
    )
    
    
    
    return df_products


## prepare data
df_products  = feature_engineering(df)
top_products = get_top_products(df)
df_products["most_revenue_product"] = df_products["product_id"].map(lambda x: x in top_products)

df_count_products = df_products["counter_by_customers"].apply(pd.Series).fillna(-1)
df_count_products.columns = ["customer_{}".format(x) for x in df_count_products.columns]

df_products_big = pd.concat([df_products, df_count_products], axis=1)

### prepare for training model: feats, X, y
feats = list(df_products_big.select_dtypes([np.number, bool]).columns)
black_list = ["most_revenue_product", "sum_price_total", "count_orders"]
feats = [x for x in feats if x not in black_list]


X = df_products_big[feats].values
y = df_products_big["most_revenue_product"].values


xgb_params = dict(
    max_depth=5, 
    n_estimators=100, 
    learning_rate=0.3, 
    objective="binary:logistic", 
    eval_metric="error",
    random_state=0)
model = xgb.XGBClassifier(**xgb_params)

#validate model
mean_score, std_score = train_and_get_scores(model, X, y)
print(f"score: mean={mean_score:.2f}, std={std_score:.2f}")


#check feature importances
model.fit(X, y)
eli5.show_weights(model, feature_names=feats, top=50)
```
Старайся заглядывать в эту секцию только после того, как самостоятельно сделаешь задачу, чтобы сверить свое решение с нашим. 😉
</p>
</details> 
</p>
</details>

In [ ]:
### YOUR CODE HERE
def get_top_products(df):
    df_products = (
        df[ ["price_total", "product_id"] ]
        .groupby("product_id")
        .agg("sum")
        .reset_index()
        .sort_values(by="price_total", ascending=False)
        .rename(columns={"price_total": "product_price_total"})
    )


    df_products["cumsum"] = df_products["product_price_total"].cumsum()
    value_80prc = int(df["price_total"].sum() * 0.8)
    df_products["most_revenue_product"] = df_products["cumsum"] < value_80prc


    top_products = set(df_products[ df_products["most_revenue_product"] ]["product_id"].unique())

    del df_products
    gc.collect()
    return top_products

def feature_engineering(df):
    
    def counter(vals):
        cntr = Counter()
        cntr.update(vals)
        return cntr
    
    df_products = (
        df
        .groupby("product_id")
        .agg(
            counter_by_customers=("customer_id", lambda x:  counter(x) ),
            sum_quantity=("quantity", np.sum),
            sum_price_unit=("price_unit", np.sum),
            sum_price_total=("price_total", np.sum),
            count_unq_countries=("country_id", lambda x: len(set(x))),
            prob_canceled=("is_canceled", np.mean)
        ).reset_index()
    )
    
    
    
    return df_products

df_products  = feature_engineering(df)
top_products = get_top_products(df)
df_products["most_revenue_product"] = df_products["product_id"].map(lambda x: x in top_products)
df_count_products = df_products["counter_by_customers"].apply(pd.Series).fillna(-1)
df_count_products.columns = ["customer_{}".format(x) for x in df_count_products.columns]
df_products_big = pd.concat([df_products, df_count_products], axis=1)
feats = list(df_products_big.select_dtypes([np.number, bool]).columns)
black_list = ["most_revenue_product", "sum_price_total", "count_orders"]
feats = [x for x in feats if x not in black_list]
X = df_products_big[feats].values
y = df_products_big["most_revenue_product"].values
xgb_params = dict(
    max_depth=5, 
    n_estimators=100, 
    learning_rate=0.3, 
    objective="binary:logistic", 
    eval_metric="error",
    random_state=0)
model = xgb.XGBClassifier(**xgb_params)
mean_score, std_score = train_and_get_scores(model, X, y)
print(f"score: mean={mean_score:.2f}, std={std_score:.2f}")

model.fit(X, y)
eli5.show_weights(model, feature_names=feats, top=50)

# Если задание выполнено, нужно сделать следующее:
1. Сделать скриншот, на котором видно решение задания
2. Добавить этот скриншот на канале [# dwthon_day5_done](https://bit.ly/3x35SfX)
3. Получить бонус за хорошую работу 💪 и завтра будет еще одна порция.

Внимание! Если у тебя возникла проблемка с задачей или чего-то не знаешь, не помнишь, не понимаешь - смело спрашивай! Это обучение :) Задавать вопросы по 5-му дню DWthon нужно в чате [# dwthon_day5](https://bit.ly/3z2qO8p).

## 🧠 Включим критическое мышление


Весь наш анализ был основан на `price_total`, что является доходом. А как насчет возвратов? Может, пора было еще немного углубиться в кроличью нору?

На что еще стоит обратить внимание? Сразу скажу, что есть еще много чего;)

## 😇 Включим творческое мышление

Подумай, как можно использовать полученные здесь знания?

В идеале есть примеры с Твоего собственного двора. Если Ты этого не сделаешь, думая, что Ты просто не сможешь этого сделать - подумай о своем окружении.

Дело не только в E-commerce. Это может быть любая отрасль. Правило 80/20 универсально. То же самое и с «aha-moment».

## 🤝🗣️ Совместное обсуждение и обмен опытом
Было бы неплохо это обсудить :)

#### ☝️ Поделись своими мыслями в Slacke **[#dwthon_day5_ideas](https://bit.ly/3ifwx4Y)**. 

Если у Тебя только идеи, но нет кода - ничего страшного, во-первых всегда можно попросить помощи в Slack. Во-вторых, важны идеи! 
Если есть идеи и код, то тем более поделись. Таким образом, кто-то другой сможет легко повторить твой опыт, вдохновиться и выдумать что-то новое и потом поделиться своими мыслями. Благодаря этому, выигрывают все :).

In [ ]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/NdqHTRJYMHU" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## 🥳🥳🥳 Поздравляю, до конца последнего урока не многие дошли - Ты молодец 💪
В Тебе есть потенциал, стремление и желание развиваться - это главное для достижения любых целей.
Чтобы не останавливаться на достигнутом (за 5 дней Тебе удалось уже много узнать 😉), сейчас приглашаю на полную версию курсов:

1️⃣ [Практическое введение в Python для Data Science](https://dataworkshop.ru/kurs/python-for-data-science) - это первый уровень обучения. С него стоит начать, если ранее Тебе не приходилось писать код на Python. В данном курсе сделан упор не только на Python, но и на Data Science.

*Немного истории:*

Курс по Python появился благодаря пожеланиям наших студентов, который приходили на курс по Data Science с нулевыми знаниями языка Python. Кстати, практически все курсы в DataWorkshop появились по просьбам наших студентов 😎.*

[Согласно индексу TIOBE](https://www.tiobe.com/tiobe-index/), Python является самым популярным языком в мире. Если планируешь получить те навыки, которые не потеряют актуальность в ближайшие годы (и даже будут пользоваться все большим спросом у работодателей) - искренне рекомендую присмотреться к этому языку программирования и начать в нем практиковаться.


2️⃣ [Курс по Data Science](https://dataworkshop.ru/kurs/data-science) - это наш главный курс, с которого и начиналась образовательная деятельность в DataWorkshop. На нем получаешь немного теории, а все остальное - практика на реальных данных, которая как раз и нужна для работы на реальных проектах!

**Важно!**
- 👉 на всех своих курсах я провожу обучение так, как хотел бы чтобы обучили в свое время меня.
- 👉 мой практический опыт в коммерческом программировании - 15 лет. В Data Science - 10+ лет. Свой опыт я получал в европейских и американских компаниях. Прошел путь от Junior разработчика до архитектора. Последнее место работы - General Electric, а затем создал свою компанию - DataWorkshop.
- 👉 наша компания занимается проектами по машинному обучению - мы практики и постоянно помогаем внедрять машинное обучение различным компаниям (агентство по недвижимости (о нашем сотрудничестве написали в польском Forbes), заводы, крупнейшая компания оператор сотовой связи, e-commerce и др). Поэтому на своих курсах я делюсь именно тем, что реально работает на практике, а не в книжках - это большая разница.
- 👉 Как видишь, машинное обучение используется в самых различных сферах и его популярность только набирает обороты. Стоит присмотреться к сфере Data Science - это очень перспективное и востребованное направление. От себя добавлю, что это еще и очень интересное направление.

Действуй, а мы поможем ❤️.

## Отзыв и дельта Твоего прогресса

Дай мне, пожалуйста, знать, что Тебе понравилось, было ли трудно (и где именно) и зафиксируй также для себя - чему получилось научиться на этом уроке.

# Тут 👉👉 [анкета](https://form.typeform.com/to/JGSWbqEK) 👈👈
Твой отзыв позволит улучшить материал!

## 💪 Покажи миру, что DWthon пройден 💪

Ты сделал(а) это! Ты на финише, обязательно расскажи об этом другим (например, в LinkedIn, Instagram 😉). Возможно, это вдохновит кого-то действовать, и, кроме того, стоит вслух рассказать о новых навыках и опыте, потому что никогда не знаешь, какие двери могут открыться;)

Помни о #dwthon #hackoutsidethebox;)

Нам очень нужна Твоя поддержка!

## И отметь, пожалуйста, нас в [LinkedIn](https://www.linkedin.com/company/dataworkhsop) и других  соц. сетях - будем рады :))